added decrypt encryption result methods

hongfeiyang opened the pull request:
Added methods to decrypt `EncryptionResult` directly instead of having to serialise it and deserialise it to decrypt
Fixed a bug related to generating a `DataEncryptionKey`, which calls `SecureRandom()` without supplying any algorithms
Changed API of `decryptWithKey()` to incorporate both decrypting serialise strings and `EncryptionResult` objects

zbarbuto commented:
I'd prefer we just keep the one consistent method name and make the parameters consistent across the board rather than adding these split methods. We are already addressing issues in other cryppo ports of the surface API getting too big and confusing. I think make all methods consistent and something like the following:

```ts
decryptWithKey({ EncryptionResult encryptionResult, String serialized, EncryptionKey key }) async {
  assert(encryptionResult != null || serialized != null,, 'Please provide an EncryptionResult or serialized encrypted string')
  assert(encryptionResult == null || serialized == null, 'Please provide either an EncryptionResult or serialized encrypted string (not both)' );
  if(encryptionResult != null) {
    // current method you already have but now private
    return _decryptEncryptionResultWithKey( ... )
  }
  if(encryptionResult != null) {
    // current method you already have but now private
    serialized _decryptSerializedResultWithKey( ... )
  }
}
```

hongfeiyang commented:
> I'd prefer we just keep the one consistent method name and make the parameters consistent across the board rather than adding these split methods. We are already addressing issues in other cryppo ports of the surface API getting too big and confusing. I think make all methods consistent and something like the following:
> 
> ```ts
> decryptWithKey({ EncryptionResult encryptionResult, String serialized, EncryptionKey key }) async {
>   assert(encryptionResult != null || serialized != null,, 'Please provide an EncryptionResult or serialized encrypted string')
>   assert(encryptionResult == null || serialized == null, 'Please provide either an EncryptionResult or serialized encrypted string (not both)' );
>   if(encryptionResult != null) {
>     // current method you already have but now private
>     return _decryptEncryptionResultWithKey( ... )
>   }
>   if(encryptionResult != null) {
>     // current method you already have but now private
>     serialized _decryptSerializedResultWithKey( ... )
>   }
> }
> ```

<img width="1450" alt="image" src="https://user-images.githubusercontent.com/19302128/104531189-b9907280-5661-11eb-899a-fe23dc89d37e.png">

The top level API did not split, only the layer beneath it was split into two

zbarbuto commented:
My mistake - haven't looked at this codebase for a while and forgot how it fit together.
